### PR TITLE
Add test for PK parsing of keys using compressed points

### LINKED ISSUE #6886: Add test for PK parsing of keys using compressed points
### Suggested enhancement

Following https://github.com/Mbed-TLS/mbedtls/pull/6282 that added support for reading compressed points in ECP, the PK parse module is supposedly able to read EC keys with compressed points in them. However, that is not tested. This task is to add tests for that.

Note: keys with compressed points can be generated with OpenSSL using `-conv_form compressed`, see the second answer to [this SO question](https://stackoverflow.com/questions/29355027/what-method-does-openssl-use-to-combine-a-public-ec-keys-coordinates).

### Justification

Mbed TLS needs this because everything that's supposed to work needs to be tested.

### PR DESCRIPTION
**Pull request by valeriosetti**
## Description
Following merging of PR #6282, `PK parse` module should now be able to parse EC compressed points. This PR adds some test for that purpose.
Resolves #6886 

## Gatekeeper checklist

- [x] **changelog** not required - tests only
- [x] **backport** not required - tests for a feature that's only in development
- [x] **tests** provided - the point of this PR

### CODE REVIEW

File: tests/data_files/Makefile
Comment (valeriosetti) [reply tests/data_files/Makefile]: Ops, you're right, of course

**Review by mpg — CHANGES_REQUESTED** [review]
What's there looks good to me, but I have a question about what isn't there: why no `ec_224_pub.comp.pem` (and associated `Parse Public EC Key #3a` test case)?

**Comment by valeriosetti** [comment]
> What's there looks good to me, but I have a question about what isn't there: why no `ec_224_pub.comp.pem` (and associated `Parse Public EC Key #3a` test case)?

Because I'm not sure those curves are supported:
https://github.com/Mbed-TLS/mbedtls/blob/development/ChangeLog.d/mbedtls_ecp_point_read_binary-compressed-fmt.txt

Indeed if I try to add it then `mbedtls_ecp_sw_derive_y` return at line [1280](https://github.com/Mbed-TLS/mbedtls/blob/8296eabed618351be21f171f5089a0f4c9c5c5e2/library/ecp.c#L1280)

**Comment by mpg** [comment]
Ah good point :) Still, it might be better to be explicit about it and add a negative test case with a comment then. Otherwise it's hard to know if the missing case is intentional or an oversight when looking at it in the future.

File: tests/suites/test_suite_pkparse.function
Comment (mpg) [inline tests/suites/test_suite_pkparse.function:16]: I think we could achieve the same effect with `#include "mbedtls/ecp.h"` as the error codes are defined unconditionally.

**Review by mpg — CHANGES_REQUESTED** [review]
Looks good except one point where I think a cleaner solution is available.

File: tests/suites/test_suite_pkparse.function
Comment (valeriosetti) [reply tests/suites/test_suite_pkparse.function]: Your solution is way more elegant than mine, of course, so I just force pushed it taking the old one away ;)

**Review by mpg — APPROVED** [review]
LGTM

File: tests/data_files/Makefile
Comment (gilles-peskine-arm) [inline tests/data_files/Makefile:867]: Non-blocker: it's generally preferable to have test files in DER format (except when testing PEM-specific features), so they can be used in build without PEM.

File: tests/suites/test_suite_pkparse.data
Comment (gilles-peskine-arm) [inline tests/suites/test_suite_pkparse.data:923]: Where is this limitation documented? I can't find anything about it in `ecp.h`.

File: tests/suites/test_suite_pkparse.data
Comment (gilles-peskine-arm) [inline tests/suites/test_suite_pkparse.data:1080]: Missing: public and private keys on Koblitz curves (192k1, 224k1, 256k1).

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]
Koblitz curves are missing. That would be ok if there was another tracked task to test them.

File: tests/data_files/Makefile
Comment (mpg) [reply tests/data_files/Makefile:867]: Good, point, though that's pre-existing, in that the files with uncompressed format are already PEM.

File: tests/suites/test_suite_pkparse.data
Comment (mpg) [reply tests/suites/test_suite_pkparse.data:923]: Good point again. I'm afraid right now it's only explicitly documented in [the ChangeLog entry](https://github.com/Mbed-TLS/mbedtls/pull/6282/files#diff-3b9d707a71bac75fb3ee1fb8f9bfce78c237752dee275586090126bee530b5e5R1) (and implicitly in the comments of `mbedtls_ecp_sw_derive_y` in `ecp.c`) which is not great.

I think we should fix that by adding something to the doxygen documentation of `mbedtls_ecp_point_read_binary()` (along the lines of the ChangeLog entry) and then ideally referencing it in its various callers. Wdyt? Where would you expect to find this documented?

**Comment by mpg** [comment]
I'd like to suggest declareing the following pre-existing issues out of scope and track them in a separate issue: (1) test files should be DER, (2) missing tests for "Koblitz" curves.

OTOH, while the documentation gap is pre-existing, I think we should fix it here. One reason is I'd like this fixed sooner rather than later. The other reason is it fits with the them of the PR: testing the consequences of compressed parsing on PK - as a general rule documentation and tests should be aligned, so this PR should update the documentation of PK parse functions, and then it might as well address the gap in `ecp.h` at the same time.

File: tests/suites/test_suite_pkparse.data
Comment (gilles-peskine-arm) [reply tests/suites/test_suite_pkparse.data:923]: I would expect this to be documented for functions whose primary goal is to parse a point or an ECC public key: `mbedtls_ecp_point_read_binary`, `mbedtls_pk_parse_public_key`, and also in the documentation of `MBEDTLS_ECP_PF_COMPRESSED`. This can be by reference (e.g. “For compressed points, see `#MBEDTLS_ECP_PF_COMPRESSED` for limitations.”)

File: include/mbedtls/ecp.h
Comment (mpg) [inline include/mbedtls/ecp.h:425]: Ah, unfortunately it's a bit more complex: _writing_ compressed format is supported for all curves, but _parsing_ it is supported as written.

Also, this comment is not doxygen, so people who consult the rendered doxygen documentation (one we have it online again) would miss it. I think we should include that in the doxygen documentation of `PF_COMPRESSED`.

Suggestion:
```c
/**
 * The uncompressed point format for Short Weierstrass curves
 * (MBEDTLS_ECP_DP_SECP* and MBEDTLS_ECP_DP_BP*).
 */
#define MBEDTLS_ECP_PF_UNCOMPRESSED    0
/**
 * The compressed point format for Short Weierstrass curves
 * (MBEDTLS_ECP_DP_SECP* and MBEDTLS_ECP_DP_BP*).
 *
 * \warning While this format is supported for all concerned curves for
 * writing, when it comes to parsing, it is not supported for all curves.
 * Specifically, parsing compressed points on MBEDTLS_ECP_DP_SECP224R1 and
 * MBEDTLS_ECP_DP_SECP224K1 is not supported.
 */
#define MBEDTLS_ECP_PF_COMPRESSED      1

```

**Review by mpg — CHANGES_REQUESTED** [review]
Just one bit of documentation to improve.

**Comment by gilles-peskine-arm** [comment]
CI is still going, but already unhappy

**Review by mpg — DISMISSED** [review]
LGTM

**Comment by mpg** [comment]
I created https://github.com/Mbed-TLS/mbedtls/issues/7060 to track the parts that we decided to consider out of scope here. @gilles-peskine-arm perhaps that should be added to some tech debt list?

Also, CI now green so this is ready for re-review.

**Comment by mpg** [comment]
Note to self & other gatekeepers: I suspect this might conflict (as is not seen by git but only by the CI) with https://github.com/Mbed-TLS/mbedtls/pull/6549 which was just merged, to we want to wait for the nightly re-run of pr-merge before merging this. Temporarily labeling "do not merge" "needs-ci" as a reminder.

**Comment by mpg** [comment]
No conflict with #6970. Now that I see the results, I should have known: 6970 is about building without ECDSA, and we'll run into problem with compressed points only when building without ECP. Anyway, better safe than sorry.

File: include/mbedtls/ecp.h
Comment (gilles-peskine-arm) [inline include/mbedtls/ecp.h:424]: This includes “//no-check-names” in the documentation. That's not good.

Our normal convention is to write `XXX` or `xxx` for a wildcard, not `*`. And `check_names.py` has an exception for names containing these.

```suggestion
 * (MBEDTLS_ECP_DP_SECP_XXX and MBEDTLS_ECP_DP_BP_XXX).
```
and likewise below.

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]
Just one problem in the documentation.

**Review by mpg — APPROVED** [review]

File: include/mbedtls/ecp.h
Comment (gilles-peskine-arm) [inline include/mbedtls/ecp.h:434]: Non-blocker:
```suggestion
 *              #MBEDTLS_ECP_DP_SECP224R1 and #MBEDTLS_ECP_DP_SECP224K1 is not
```

(Doxygen automatically makes links for `foo()`, but not for a macro name.)

**Review by gilles-peskine-arm — APPROVED** [review]
LGTM

**Pull request merged**
